### ISSUE TITLE
[Accessibility Docs] Contextual menu

### ISSUE THREAD

**Issue by sowasred2012** [canonical-web-and-design/vanilla-framework#4042]
Using [this template](https://docs.google.com/document/d/1IwFMfJrTfJPXDcVJfLWXYkw-pjgWhBWo4GX8X098Q64/edit#), create accessibility docs for [Contextual menu](https://vanillframework.io/docs/patterns/contextual-menu)

**Comment by bethcollins92** [comment]
[Accessibility doc](https://docs.google.com/document/d/1uRHo5Jp034T31E18IuTSdgpzLaF77lOFGHzn7U49hqk/edit#) for contextual menu. 

I've listened to VoiceOver, the focus isn't trapped in the contextual menu and it feels like it should be? Also, when you tab off the menu, voiceover speaks "menu closed" when it doesn't actually close.

**Comment by marinacastejon** [comment]
I did the same test with VoiceOver but in iOS. I checked in Safari and Chrome, and I got the same behaviour. VO reads the following: 
When collapsed: take action, button, menu pop-up, collapsed, double-tap to expand. 
When expanded: take action, menu pop-up expanded
Then it keeps reading all the items in the menu as buttons and then it keeps reading the other elements in the page without closing the menu. 
![IMG_0050](https://user-images.githubusercontent.com/89836430/151333898-c51dad03-39e9-4c20-8dca-eb979868296e.PNG)

Regarding focus trap, here is my analysis: 

[**WCAG 2.1.2. No keyboard trap** ](https://www.w3.org/TR/WCAG21/#no-keyboard-trap)
Looking at this it doesn't seem to be a good idea to trap the focus, and that if we do choose to, we would need some extra explanation when the component is open to tell the user how to move the focus away from the component.
[Understanding the guideline with examples](https://www.w3.org/WAI/WCAG21/Understanding/no-keyboard-trap.html):
> If keyboard focus can be moved to a component of the page using a keyboard interface, then focus can be moved away from that component using only a keyboard interface, and, if it requires more than unmodified arrow or tab keys or other standard exit methods, the user is advised of the method for moving focus away.

On the other hand, the reason why the current solution is not optimal either is that the focus will not be necessarily visible at all times, assuming there will be content hidden behind the expanded menu, violating the guideline of [Focus visible](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html).

**Carbon Design system**, they call it [Overflow menu](https://www.carbondesignsystem.com/components/overflow-menu/accessibility), but I think technically they are very similar (the use ARIA like in our menu). They do not mention focus trapping, although in their live demo using keyboard only, when you tab outside the menu closes. The problem I see with this solution is that the screen reader does not announce that the pop-up menu has been closed, it only reads "list end" on the last item.

[**Yale University, Focus and Keyboard Operability**](https://usability.yale.edu/web-accessibility/articles/focus-keyboard-operability)
My understanding is that they recommend the same solution as what is implemented in the Carbon design system: 
> For complex widgets with many potential focusable elements, it may be preferable to use the Tab key to move focus into and outside the widget, and use arrow keys and scripting to manipulate focus within the element. This concept is known as roving tabindex. Developers implementing complex widgets that make use of arrow keys should consult the ARIA 1.1 Authoring Practices document’s section on keyboard navigation inside components(link is external).

[**WebAIM**](https://webaim.org/techniques/keyboard/) does not give any examples similar to our menu, but the guidelines are clear. Just as FYI.

So I went to WAI-ARIA again, and reading more thoroughly I found something interesting. The menu button points to the menu section for explanations on keyboard behaviour: 
> Tab: Moves focus to the next element in the tab sequence, and if the item that had focus is not in a menubar, closes its menu and all open parent menu containers.

So I would say as a conclusion:
When closed, on tab, focus moves away from the menu (obvious).
When closed, enter or space keys open the menu. On tab, there are 2 options:
- We close the menu and go to the next focusable element outside the menu.
- We can tab through the different elements in the menu and after the last element in the menu, on tab, the menu closes and focus goes to the next focusable element outside the menu.

**Comment by sowasred2012** [comment]
I think this should be an issue on Vanilla, as you say.

I tested https://vanillaframework.io/docs/examples/patterns/contextual-menu/default in Orca on Ubuntu and what I got was:

- Keyboard interaction between the button with a contextual menu and a link with a contextual menu isn't the same. Pressing Enter opens both, but pressing Space opens only the button example. I actually wonder if all our contextual menu CTAs should be buttons, since semantically that's what they are, they just _look_ like links in some cases.
- It does feel wrong that the menu doesn't close when you tab off it. I don't think we need to trap the focus.

**Comment by bethcollins92** [comment]
Ok, after reading all this (thank you so much!) I'll block this accessibility doc issue by an [issue](https://github.com/canonical-web-and-design/vanilla-framework/issues/4284) to fix the menu closing when the focus leaves.

**Comment by marinacastejon** [comment]
@bethcollins92 @sowasred2012 in the keyboard interaction guidelines we still decide if tab moves to the next element inside the menu or if it moves out of the menu and closes it (meaning you can navigate inside the menu with arrow keys, ala Carbon Design System). Usability-wise I don't believe there is a better option, you may decide what makes more sense technically speaking.

**Comment by bethcollins92** [comment]
The contexrtual menu was [fixed](https://github.com/canonical-web-and-design/vanilla-framework/pull/4359) so this doc can now be finished and added to Vanilla
